src/utils/index.ts

```diff
@@ -248,6 +264,150 @@ export interface ICurrentStats {
   };
 }
 
+export interface ICirculatingSupply {
+  heldByMultisig: BigNumber;
+  heldByUBQINK: BigNumber;
+  heldByGRANSUBQ: BigNumber;
+  heldByUBQESCH: BigNumber;
+  total: BigNumber;
+}
+
+export interface IDaoHoldings {
+  ubq: BigNumber;
+  ink: BigNumber;
+  esch: BigNumber;
+  lp: {
+    ubqInk: BigNumber;
+    gransInk: BigNumber;
+    inkEsch: BigNumber;
+  };
+}
+
+export const getDaoHoldings = async (provider: provider): Promise<IDaoHoldings> => {
+  // Get coin and token holdings from the DAO multisig address and farming address
+  const ubqHoldings = (await getCoinBalanceAsBigNum(provider, DAO_MULTISIG)).plus(await getCoinBalanceAsBigNum(provider, DAO_FARMING));
+  const inkHoldings = (await getBalanceAsBigNum(provider, INK, DAO_MULTISIG)).plus(await getBalanceAsBigNum(provider, INK, DAO_FARMING));
+  const eschHoldings = (await getBalanceAsBigNum(provider, ESCH, DAO_MULTISIG)).plus(await getBalanceAsBigNum(provider, ESCH, DAO_FARMING));
+
+  // Get lp holdings from the DAO multisig address and farming address
+  const ubqInkHoldings = (await getBalanceAsBigNum(provider, INK_UBQ_LPAddress, DAO_MULTISIG)).plus(
+    await getBalanceAsBigNum(provider, INK_UBQ_LPAddress, DAO_FARMING)
+  );
+  const gransInkHoldings = (await getBalanceAsBigNum(provider, INK_GRANS_LPAddress, DAO_MULTISIG)).plus(
+    await getBalanceAsBigNum(provider, INK_GRANS_LPAddress, DAO_FARMING)
+  );
+  const inkEschHoldings = (await getBalanceAsBigNum(provider, INK_ESCH_LPAddress, DAO_MULTISIG)).plus(
+    await getBalanceAsBigNum(provider, INK_ESCH_LPAddress, DAO_FARMING)
+  );
+
+  return {
+    ubq: ubqHoldings,
+    ink: inkHoldings,
+    esch: eschHoldings,
+    lp: {
+      ubqInk: ubqInkHoldings,
+      gransInk: gransInkHoldings,
+      inkEsch: inkEschHoldings,
+    },
+  } as IDaoHoldings;
+};
+
+// Calculates the circulating INK supply based on the amount still held by
+// the DAO minting address plus a sum of all of the individual farm holdings
+export const getCirculatingSupply = async (provider: provider): Promise<ICirculatingSupply> => {
```

Comment:
is there a check somewhere to ensure, at runtime, that the provider is set?

Reply:
it doesn't look like they've ever done that in this file, so i don't know how that works